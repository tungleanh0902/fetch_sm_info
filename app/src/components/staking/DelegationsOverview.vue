<template>
  <div class="delegations-overview">
    <div class="table-container">
      <div
        v-if="$apollo.queries.delegations.loading && !delegationsLoaded"
        class="loading-image-container"
      >
        <img
          class="loading-image"
          src="/img/portfolio-loading.svg"
          alt="geometric placeholder shapes"
        />
      </div>
      <div v-else-if="delegations.length > 0 || stakedBalance.total > 0">
        <h1>Your Stake</h1>
        <BalanceRow
          v-if="stakedBalance.total > 0"
          :balance="stakedBalance"
          :total-rewards-per-denom="totalRewardsPerDenom"
          :stake="currentNetwork.network_type === 'polkadot'"
          :unstake="currentNetwork.network_type === 'polkadot'"
        />
        <TableValidators
          v-if="delegations.length > 0"
          :validators="delegations.map(({ validator }) => validator)"
          :delegations="delegations"
          class="table-validators"
          show-on-mobile="expectedReturns"
        />
      </div>
      <TmDataMsg
        v-if="!$apollo.loading && delegations.length === 0"
        icon="sentiment_dissatisfied"
      >
        <div slot="title">No validators in your portfolio</div>
        <div slot="subtitle">
          Head over to the
          <a @click="goToValidators()">validator list</a>&nbsp;to
          {{
            stakedBalance.total > 0 ? `start earning rewards` : `get staking`
          }}!
        </div>
      </TmDataMsg>
      <UndelegationModal ref="UnstakeModal" />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState } from "vuex"
import BalanceRow from "common/BalanceRow"
import TmDataMsg from "common/TmDataMsg"
import TableValidators from "staking/TableValidators"
import UndelegationModal from "src/ActionModal/components/UndelegationModal"
import {
  ValidatorFragment,
  DelegationsForDelegator,
  UserTransactionAdded,
} from "src/gql"
import gql from "graphql-tag"

export default {
  name: `delegations-overview`,
  components: {
    BalanceRow,
    TableValidators,
    UndelegationModal,
    TmDataMsg,
  },
  data: () => ({
    delegations: [],
    balances: [],
    rewards: [],
    delegationsLoaded: false,
    preferredCurrency: "USD",
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`address`, `currentNetwork`]),
    stakedBalance() {
      // balances not loaded yet
      if (!this.balances.length) {
        return {
          total: 0,
          denom: this.currentNetwork.stakingDenom,
        }
      }
      const stakingDenomBalance = this.balances.find(
        ({ denom }) => denom === this.currentNetwork.stakingDenom
      )
      return {
        total: stakingDenomBalance
          ? Number(stakingDenomBalance.staked).toFixed(3)
          : 0,
        denom: this.currentNetwork.stakingDenom,
      }
    },
    totalRewardsPerDenom() {
      return this.rewards.reduce((all, reward) => {
        return {
          ...all,
          [reward.denom]: parseFloat(reward.amount) + (all[reward.denom] || 0),
        }
      }, {})
    },
  },
  mounted: function () {
    const persistedPreferredCurrency = this.session.preferredCurrency
    if (persistedPreferredCurrency) {
      this.preferredCurrency = persistedPreferredCurrency
    }
  },
  methods: {
    goToValidators() {
      this.$router.push({
        name: "validators",
        params: {
          networkId: this.currentNetwork.slug,
        },
      })
    },
    openUnstakeModal() {
      this.$refs.UnstakeModal.open()
    },
  },
  apollo: {
    balances: {
      query: gql`
        query($networkId: String!, $address: String!) {
          balancesV2(networkId: $networkId, address: $address) {
            id
            type
            denom
            available
            total
            staked
          }
        }
      `,
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.currentNetwork.id,
          address: this.address,
        }
      },
      /* istanbul ignore next */
      skip() {
        return !this.address
      },
      update(result) {
        return result.balancesV2
      },
    },
    delegations: {
      query() {
        /* istanbul ignore next */
        return DelegationsForDelegator(this.currentNetwork.id)
      },
      variables() {
        /* istanbul ignore next */
        return {
          delegatorAddress: this.address,
          networkId: this.currentNetwork.id,
        }
      },
      /* istanbul ignore next */
      update(data) {
        this.delegationsLoaded = true
        return data.delegations || []
      },
    },
    rewards: {
      query: gql`
        query rewards(
          $networkId: String!
          $delegatorAddress: String!
          $fiatCurrency: String
        ) {
          rewards(
            networkId: $networkId
            delegatorAddress: $delegatorAddress
            fiatCurrency: $fiatCurrency
          ) {
            id
            amount
            denom
          }
        }
      `,
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.currentNetwork.id,
          delegatorAddress: this.address,
          fiatCurrency: this.preferredCurrency,
        }
      },
      /* istanbul ignore next */
      skip() {
        return !this.address
      },
    },
    $subscribe: {
      userTransactionAdded: {
        variables() {
          return {
            networkId: this.currentNetwork.id,
            address: this.address,
          }
        },
        skip() {
          return !this.address
        },
        query: UserTransactionAdded,
        result({ data }) {
          /* istanbul ignore next */
          if (data.userTransactionAddedV2.success) {
            this.$apollo.queries.delegations.refetch()
          }
        },
      },
    },
  },
}
</script>
<style scoped>
h1 {
  font-size: 24px;
  color: var(--bright);
  font-weight: 400;
  padding-bottom: 2rem;
}

.delegations-overview {
  background: var(--app-fg);
}

.table-container {
  max-width: 1100px;
  margin: 0 auto;
  width: 100%;
  padding: 4rem 2rem;
}

.table-validators {
  margin-top: 2rem;
}

.tm-form-msg--desc {
  padding-bottom: 1rem;
}

.tm-data-msg {
  margin-top: 1rem;
}

@media screen and (max-width: 667px) {
  h1 {
    padding: 2rem;
    text-align: center;
  }

  .loading-image-container {
    padding: 2rem;
  }

  .table-container {
    padding: 4rem 1rem;
  }
}
</style>
