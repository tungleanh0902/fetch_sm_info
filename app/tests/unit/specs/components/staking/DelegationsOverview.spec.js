import { shallowMount, createLocalVue } from "@vue/test-utils"
import DelegationsOverview from "staking/DelegationsOverview"
import validators from "../../store/json/validators.js"
import Vuex from "vuex"

describe(`DelegationsOverview`, () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)

  let wrapper, $store, $apollo, delegations

  const getters = {
    committedDelegations: {
      [validators[0].operator_address]: validators[0],
    },
    address: "cosmos1",
    currentNetwork: {
      id: "testnet",
      stakingDenom: "STAKE",
    },
  }

  delegations = [
    {
      validator: validators[0],
      delegatorAddress: `cosmos1`,
      amount: 10,
    },
    {
      validator: validators[1],
      delegatorAddress: `cosmos1`,
      amount: 12,
    },
    {
      validator: validators[2],
      delegatorAddress: `cosmos1`,
      amount: 11,
    },
  ]

  beforeEach(() => {
    $store = {
      getters,
      state: {
        connection: {
          network: "testnet",
        },
        session: {
          addressRole: undefined,
        },
      },
      address: "cosmos1",
      network: "testnet",
    }

    $apollo = {
      queries: {
        delegations: {
          loading: false,
          error: false,
        },
        balances: {
          loading: false,
          error: false,
        },
      },
    }

    wrapper = shallowMount(DelegationsOverview, {
      mocks: {
        $store,
        $apollo,
      },
      stubs: [`router-link`],
    })

    wrapper.setData({
      delegations,
      balances: [
        {
          total: 10000,
          liquid: 50,
          staked: 9850, // 100 would be unstaking
          denom: "STAKE",
        },
      ],
    })
  })

  it(`shows an overview over all delegations of the user`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`shows a sentiment of dissatisfaction when you have no such delegations`, async () => {
    wrapper.setData({
      delegations: [],
      balances: [
        {
          total: 34,
          available: 1,
          staked: 33,
          denom: "STAKE",
        },
      ],
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
