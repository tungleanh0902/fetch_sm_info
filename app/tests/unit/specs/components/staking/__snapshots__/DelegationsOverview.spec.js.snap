// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelegationsOverview shows a sentiment of dissatisfaction when you have no such delegations 1`] = `
<div
  class="delegations-overview"
>
  <div
    class="table-container"
  >
    <div>
      <h1>
        Your Stake
      </h1>
       
      <balancerow-stub
        balance="[object Object]"
        totalrewardsperdenom="[object Object]"
      />
       
      <!---->
    </div>
     
    <tmdatamsg-stub
      icon="sentiment_dissatisfied"
      iconcolor="var(--warning)"
    >
      <div>
        No validators in your portfolio
      </div>
       
      <div>
        
        Head over to the
        
        <a>
          validator list
        </a>
        Â to
        start earning rewards!
      
      </div>
    </tmdatamsg-stub>
     
    <undelegationmodal-stub
      sourcevalidator="[object Object]"
    />
  </div>
</div>
`;

exports[`DelegationsOverview shows an overview over all delegations of the user 1`] = `
<div
  class="delegations-overview"
>
  <div
    class="table-container"
  >
    <div>
      <h1>
        Your Stake
      </h1>
       
      <balancerow-stub
        balance="[object Object]"
        totalrewardsperdenom="[object Object]"
      />
       
      <tablevalidators-stub
        class="table-validators"
        delegations="[object Object],[object Object],[object Object]"
        showonmobile="expectedReturns"
        validators="[object Object],[object Object],[object Object]"
      />
    </div>
     
    <!---->
     
    <undelegationmodal-stub
      sourcevalidator="[object Object]"
    />
  </div>
</div>
`;
